File: JAVA/S14_BST/L1_CONCEPT/_3_minmax.java
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data=data;
        left=null;
        right=null;
    }
}
class BST{
    Node root;
    Node insertbst(Node root,int data){
        Node newnode=new Node(data);
        if(root==null){
            // Node newnode=new Node(data);
            root=newnode;
            return root;
        }
        if(root.data>data){
            root.left=insertbst(root.left,data);
        }
        else root.right=insertbst(root.right,data);
        return root;

    }
    Node minValue(Node root){
        Node temp=root;
        while(temp.left!=null){
            temp=temp.left;
        }
        return temp;
    }
    Node maxValue(Node root){
        Node temp=root;
        while(temp.right!=null){
            temp=temp.right;
        }
        return temp;
    }
    void postorder(Node root){
        if(root==null)return;
        postorder(root.left);
        postorder(root.right);
        System.out.print(root.data+" ");
    }
}
public class _3_minmax{
    public static void main(String[] args){
        BST b=new BST();
        b.root=b.insertbst(b.root,67);
        b.root=b.insertbst(b.root,69);
        b.root=b.insertbst(b.root,62);
        b.root=b.insertbst(b.root,601);
        Node min=b.minValue(b.root);
        Node max=b.maxValue(b.root);
        System.out.println(min.data);
        System.out.println(max.data);
        b.postorder(b.root);
    }
}

File: python/S14_BST/L1_CONCEPT/3_minmax.py
class Node:
    def __init__(self,data):
        self.left=None
        self.right=None
        self.data=data
class bst:
    def __init__(self):
        self.root=None
    def insertbst(self,root,data):
        if root is None:
            return Node(data)
        if root.data>data:
            root.left=self.insertbst(root.left,data)
        elif root.data<data:
            root.right=self.insertbst(root.right,data)
        return root
        
            
    def maximum(self,root):
        if root is None:
            return None
        temp=root
        while temp.right is not None:
            temp=temp.right
        return temp
    def minimum(self,root):
        if root is None:
            return None
        temp=root
        while temp.left is not None:
            temp=temp.left
        return temp
    def postorder(self,root):
        if root is None:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.data,end=" ")
if __name__=="__main__":
    b=bst()
    b.root=b.insertbst(b.root,34)
    b.root=b.insertbst(b.root,30)
    b.root=b.insertbst(b.root,35)
    b.root=b.insertbst(b.root,33)
    b.root=b.insertbst(b.root,32)
    min=b.minimum(b.root)
    max=b.maximum(b.root)
    print(min.data)
    print(max.data)
    b.postorder(b.root)
    